core/src/main/scala/za/co/absa/commons/S3Location.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.commons
+
+import scala.util.matching.Regex
+
+object S3Location {
+
+  /**
+   * Generally usable regex for validating S3 path, e.g. `s3://my-cool-bucket1/path/to/file/on/s3.txt`
+   * Protocols `s3`, `s3n`, and `s3a` are allowed.
+   * Bucket naming rules defined at [[https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html#bucketnamingrules]] are instilled.
+   */
+  val S3LocationRx: Regex = "(s3[an]?)://([-a-z0-9.]{3,63})/(.*)".r
+
+  implicit class StringS3LocationExt(path: String) {
+
+    def toS3Location: Option[SimpleS3Location] = {
+      path match {
+        case S3LocationRx(protocol, bucketName, relativePath) => Some(SimpleS3Location(protocol, bucketName, relativePath))
+        case _ => None
+      }
+    }
+
+    def toS3LocationOrFail: SimpleS3Location = {
```

Comment:
Why do you need both `toS3Location` and `toS3LocationOrFail` methods?
If it's only about different code styles then it could be worth it if it were some kind or DSL implementation, but is it really worth it in this case? (Likewise you wouldn't add extra methods here returning `Try` or `Either` that only differs in return type, would you?)

What about only leaving the one that returns `Option`? The caller could always switch to any other alternatives in the place of invocation.

Reply:
You are absolutely right, I am (as outline above) aiming for a general-purpose usage. The extra stuff here, we don't need, will remove.